src/main/java/net/minecraftforge/client/gui/config/ControlCreator.java

```diff
@@ -0,0 +1,525 @@
+        T defaultValue = (T) interactor.getData(DEFAULT_VALUE_KEY);
+        T copiedDefault = defaultValue == null ? null : copyMutable(defaultValue);
+
+        interactor.label = null;
+        interactor.control = createButton(interactor, button -> {
+            Screen currentScreen = Minecraft.getInstance().currentScreen;
+            Screen screen = new ListConfigScreen(currentScreen, interactor.title, state.getValue()) {
+                @Override
+                public void onModified(List newValueIn) {
+                    T newValue = (T) newValueIn;
+                    state.setValue(newValue);
+                    boolean isValid = interactor.isValid(newValue);
+                    if (isValid)
+                        interactor.save(newValue);
+                    interactor.onUpdate(isValid, newValue);
+                }
+
+                @Override
+                protected void undo() {
+                    T newValue = copyMutable(copiedInitial);
+                    setupWithNewValue(newValue);
+                }
+
+                @Override
+                protected boolean canUndo() {
+                    return !Objects.equals(value, copiedInitial);
+                }
+
+                @Override
+                protected void reset() {
+                    if (copiedDefault == null)
+                        return;
+                    T newValue = copyMutable(copiedDefault);
+                    setupWithNewValue(newValue);
+                }
+
+                @Override
+                protected boolean canReset() {
+                    return copiedDefault != null && !Objects.equals(value, copiedDefault);
+                }
+
+                private void setupWithNewValue(T newValue) {
+                    this.value = newValue;
+                    interactor.onUpdate(newValue);
+                    // Completely recreate everything on this screen
+                    // this.init(minecraft, scaledWidth, scaledHeight);
+                    this.func_231158_b_(field_230706_i_, field_230708_k_, field_230709_l_);
+                }
+            };
+            Minecraft.getInstance().displayGuiScreen(screen);
+        });
+
+        interactor.addUpdateResponder((isValid, newValue) -> {
+            state.setValue(newValue);
+            Widget control = interactor.control;
+            if (!isValid)
+                control.setFGColor(RED);
+            else
+                control.clearFGColor();
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> T copyMutable(T o) {
+        // Immutable objects can be returned as is
+        if (o instanceof Boolean || o instanceof Number || o instanceof String || o instanceof Enum<?>)
+            return o;
+        if (o instanceof List<?>)
+            return (T) new ArrayList<>((List<?>) o);
+        if (o instanceof UnmodifiableConfig)
+            return (T) ((AbstractConfig) o).clone();
+        return o;
```

Comment:
Should we log something here, or are there legitimate cases where the code flow could reach here?

Reply:
I’m not sure. I figured it was better to fail silently and have possible weird behaviour with mutables than fail causing the game to crash when the config screen was opened or logging a message that would always be ignored.